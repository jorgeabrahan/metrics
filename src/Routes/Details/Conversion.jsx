/* eslint-disable import/no-extraneous-dependencies */
import PropTypes from 'prop-types';

const Conversion = ({ name, value, isDarker }) => (
  <div className={`conversion ${isDarker ? '' : 'conversion--ruber'}`}>
    <h3 className="conversion__name">{name}</h3>
    <p className="conversion__value">{value}</p>
  </div>
);

Conversion.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.number.isRequired,
  isDarker: PropTypes.bool.isRequired,
};

export default Conversion;
