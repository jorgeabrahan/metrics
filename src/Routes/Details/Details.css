.conversions {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
}

.conversion {
  background-color: var(--dark-pink);
  padding: 1rem 0.5rem;
}

.conversion__name {
  text-transform: uppercase;
  font-weight: 600;
  font-size: 1.3rem;
  line-height: 1.4rem;
}

.conversion__value {
  font-size: 1rem;
}

.conversion--ruber {
  background-color: var(--ruber);
}

.amount-form {
  display: flex;
  align-items: center;
}

.amount, .amount__submit {
  padding: 0.3rem;
}

.amount {
  border: none;
  outline: none;
  background: none;
  background-color: var(--ruber);
  color: white;
  width: 5rem;
  text-align: center;
  margin-right: 0.5rem;
  border-radius: 0 5px 5px 0;
}

.amount__submit {
  border: none;
  outline: none;
  background: none;
  color: white;
  background-color: #6a0b2e;
  border-radius: 5px 0 0 5px;
}