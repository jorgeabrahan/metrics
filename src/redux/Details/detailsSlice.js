/* eslint-disable import/no-extraneous-dependencies */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const API_DATE = 'latest';
const API_VERSION = '1';
const BASE_URL = `https://cdn.jsdelivr.net/gh/fawazahmed0/currency-api@${API_VERSION}/${API_DATE}/currencies`;

const initialState = {
  details: {},
  filteredConversions: [],
  amount: 1,
  status: 'idle',
  error: '',
};

export const fetchCurrency = createAsyncThunk('currency/get', (currency) => (
  new Promise((resolve, reject) => {
    axios.get(`${BASE_URL}/${currency}.json`)
      .then(({ data }) => {
        const details = {
          date: data.date,
          conversions: [],
        };
        Object.entries(data).forEach((arr) => {
          const [key, value] = arr;
          if (key !== 'date') {
            details.short = key;
            Object.entries(value).forEach(([currency, equivalent]) => {
              details.conversions.push({
                name: currency,
                value: equivalent,
              });
            });
          }
        });
        resolve(details);
      })
      .catch((error) => {
        reject(error);
      });
  })
));

const detailsSlice = createSlice({
  name: 'details',
  initialState,
  reducers: {
    clearDetails: (state) => ({
      ...state, status: 'idle', details: [], filteredConversions: [], amount: 1,
    }),
    searchForConversion: (state, { payload }) => {
      const filteredConversions = state.details.conversions.filter(({ name }) => (
        name.toLowerCase().includes(payload)
      ));

      return {
        ...state,
        filteredConversions: filteredConversions.map((c) => ({
          ...c, value: Number.parseFloat(c.value * state.amount).toFixed(2),
        })),
      };
    },
    setAmount: (state, { payload }) => ({
      ...state, amount: Number.isNaN(Number(payload)) ? 1 : Number(payload),
    }),
  },
  extraReducers(builder) {
    builder
      .addCase(fetchCurrency.pending, (state) => ({
        ...state, status: 'loading',
      }))
      .addCase(fetchCurrency.fulfilled, (state, { payload }) => ({
        ...state, status: 'fulfilled', details: payload, filteredConversions: payload.conversions,
      }))
      .addCase(fetchCurrency.rejected, (state, error) => ({
        ...state, status: 'rejected', error: error.message,
      }));
  },
});

export const {
  clearDetails, searchForConversion, setAmount,
} = detailsSlice.actions;

export default detailsSlice.reducer;
