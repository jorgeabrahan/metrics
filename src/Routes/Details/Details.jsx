/* eslint-disable import/no-extraneous-dependencies */
import { useParams } from 'react-router-dom';
import { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuid } from 'uuid';
import {
  clearDetails, fetchCurrency, searchForConversion, setAmount,
} from '../../redux/Details/detailsSlice';
import Conversion from './Conversion';
import './Details.css';
import { setQuery } from '../../redux/Search/searchSlice';

/* code to determine if the Conversion element is darker according to the pattern */
const pattern = [false, true];

const resetPattern = () => {
  pattern[0] = false;
  pattern[1] = true;
};

const isDarker = (index, isLast) => {
  if (index < pattern.length) return pattern[index];
  const [first, last] = pattern;
  if (first === last) {
    pattern[1] = !last;
    if (isLast) resetPattern();
    return !last;
  }
  pattern[0] = !first;
  if (isLast) resetPattern();
  return last;
};
/* ------------------------------------------------------------------------------ */

const Details = () => {
  const dispatch = useDispatch();
  const { currency } = useParams();
  const {
    status, details, filteredConversions, error, amount,
  } = useSelector((store) => store.details);
  const { isSearching, query } = useSelector((store) => store.search);

  const onSubmit = (e) => {
    e.preventDefault();
    // se verifica que sea un numero
    const num = Number(amount);
    if (Number.isNaN(num)) return;
    // se verifica que sea positivo
    if (num < 1) return;

    dispatch(searchForConversion(query.toLowerCase()));
  };

  useEffect(() => {
    if (status !== 'idle') return undefined;
    dispatch(fetchCurrency(currency));

    return () => {
      dispatch(clearDetails());
    };
  }, [dispatch]);

  useEffect(() => {
    if (isSearching) return;
    if (filteredConversions.length === 0 && isSearching) return;
    if (status !== 'fulfilled') return;
    dispatch(searchForConversion(''));
    dispatch(setQuery(''));
  }, [dispatch, isSearching]);

  let render = <p>Loading...</p>;
  if (status === 'rejected') {
    render = (
      <p>
        [Error]:
        {error}
      </p>
    );
  }

  const filteredConversionsList = useMemo(() => (
    filteredConversions.map(({ name, value }, index, arr) => (
      <Conversion
        key={uuid()}
        name={name}
        value={Number(value)}
        isDarker={isDarker(index, index === (arr.length - 1))}
      />
    ))
  ), [filteredConversions]);

  return (
    <div>
      {status === 'fulfilled' ? (
        <>
          <div className="header">
            <form className="amount-form" onSubmit={onSubmit}>
              <button type="submit" className="amount__submit">
                <span className="material-symbols-outlined">
                  autorenew
                </span>
              </button>
              <input
                className="amount"
                type="text"
                value={amount}
                onChange={({ target }) => dispatch(setAmount(target.value))}
                onBlur={({ target }) => Number(target.value) === 0 && dispatch(setAmount(1))}
              />
              <h2>
                {' '}
                {details.short}
                {' '}
                rates
              </h2>
            </form>
            <small>{details.date}</small>
          </div>
          <small className="subtitle-separator">List of conversions</small>
          <div className="conversions">
            {filteredConversionsList}
          </div>
        </>
      ) : render}
    </div>
  );
};

export default Details;
