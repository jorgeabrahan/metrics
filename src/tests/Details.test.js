/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import axios from 'axios';
import wrap from './testResources';
import Details from '../Routes/Details/Details';

jest.mock('axios');
const mockResponse = {
  data: {
    date: '2023-03-14',
    usd: {
      aave: 0.013188,
      ada: 2.906034,
      aed: 3.673007,
      afn: 88.499179,
    },
  },
};

describe('Tests for <Details /> component', () => {
  axios.get.mockResolvedValue(mockResponse);
  render(wrap(<Details />));
  it('Should render conversions from API with values fixed to 2', () => {
    expect(screen.getByText('aave')).not.toBeNull();
    expect(screen.getByText('ada')).not.toBeNull();
    expect(screen.getByText('aed')).not.toBeNull();
    expect(screen.getByText('afn')).not.toBeNull();

    expect(screen.getByText('0.01')).not.toBeNull();
    expect(screen.getByText('2.91')).not.toBeNull();
    expect(screen.getByText('3.67')).not.toBeNull();
    expect(screen.getByText('88.5')).not.toBeNull();
  });

  it('Should not display the loading message', () => {
    expect(() => screen.getByText('Loading...')).toThrow();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrap(<Details />)).toJSON()).toMatchSnapshot();
  });
});
