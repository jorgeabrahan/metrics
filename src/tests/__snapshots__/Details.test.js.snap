// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Details /> component Should match the snapshot 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
